File: backend/remove_bg.py
import os
import json
import requests
from io import BytesIO
from PIL import Image

try:
    from rembg import remove
except ImportError:
    print("\n[!] Please install rembg first by running:")
    print("    pip install rembg pillow requests tqdm\n")
    exit(1)

try:
    from tqdm import tqdm
    HAS_TQDM = True
except ImportError:
    HAS_TQDM = False

# Path configuration
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
JSON_PATH = os.path.join(BASE_DIR, "bags.json")
OUTPUT_DIR = os.path.join(BASE_DIR, "..", "public", "bags")

def main():
    # Load JSON data
    if not os.path.exists(JSON_PATH):
        print(f"Error: JSON file not found at {JSON_PATH}")
        return

    with open(JSON_PATH, "r", encoding="utf-8") as f:
        data = json.load(f)

    # Ensure output directory exists
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    print(f"Loaded {len(data)} items from {JSON_PATH}")
    print("Starting background removal...")

    modified_count = 0
    iterator = enumerate(data)
    if HAS_TQDM:
        iterator = tqdm(iterator, total=len(data), desc="Processing images")

    for i, item in iterator:
        image_url = item.get("image_url", "")
        
        # Check if the image is remote (starts with http)
        if image_url.startswith("http"):
            # Generate a local name for the transparent image
            clean_title = "".join(c if c.isalnum() else "_" for c in item["title"][:20]).strip("_")
            filename = f"bag_{i}_{clean_title.lower()}.png"
            local_path = os.path.join(OUTPUT_DIR, filename)
            
            # Skip if already downloaded and processed
            if os.path.exists(local_path):
                # Update URL in JSON
                item["image_url"] = f"/bags/{filename}"
                continue

            try:
                # 1. Download image
                response = requests.get(image_url, timeout=15)
                if response.status_code != 200:
                    print(f"\n[!] Failed to download {image_url}")
                    continue
                
                # 2. Open image with PIL
                img = Image.open(BytesIO(response.content))
                
                # 3. Remove background locally (retains original resolution)
                output_img = remove(img)
                
                # 4. Save as transparent PNG
                output_img.save(local_path, "PNG")
                
                # 5. Update the path in the JSON to point to public assets
                item["image_url"] = f"/bags/{filename}"
                modified_count += 1
                
                if not HAS_TQDM:
                    print(f"Processed: {item['title'][:30]}... -> {filename}")

            except Exception as e:
                print(f"\n[!] Error processing item {i} ({item['title'][:30]}): {e}")

    # Save the updated JSON back
    if modified_count > 0 or any(not item.get("image_url", "").startswith("http") for item in data):
        with open(JSON_PATH, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"\n[+] Success! Updated {modified_count} image paths in bags.json.")
        print(f"[+] All transparent images are saved under: public/bags/")
    else:
        print("\n[-] No images were processed.")

if __name__ == "__main__":
    main()

File: backend/remove_bg.js
const fs = require("fs");
const path = require("path");

async function run() {
  let removeBackground;
  try {
    const module = require("@imgly/background-removal-node");
    removeBackground = module.removeBackground;
  } catch (err) {
    console.error("\n[!] Please install the required package first by running:");
    console.error("    npm install @imgly/background-removal-node\n");
    process.exit(1);
  }

  const JSON_PATH = path.join(__dirname, "bags.json");
  const OUTPUT_DIR = path.join(__dirname, "..", "public", "bags");

  if (!fs.existsSync(JSON_PATH)) {
    console.error(`Error: JSON file not found at ${JSON_PATH}`);
    return;
  }

  // Load JSON
  const data = JSON.parse(fs.readFileSync(JSON_PATH, "utf8"));
  console.log(`Loaded ${data.length} items from bags.json`);

  // Ensure output directory exists
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }

  let modifiedCount = 0;

  for (let i = 0; i < data.length; i++) {
    const item = data[i];
    const imageUrl = item.image_url || "";

    if (imageUrl.startsWith("http")) {
      const cleanTitle = item.title
        .toLowerCase()
        .replace(/[^a-z0-9]/g, "_")
        .substring(0, 20)
        .replace(/^_+|_+$/g, "");
      
      const filename = `bag_${i}_${cleanTitle}.png`;
      const localPath = path.join(OUTPUT_DIR, filename);

      if (fs.existsSync(localPath)) {
        // Skip if already processed, but update path in JSON
        item.image_url = `/bags/${filename}`;
        continue;
      }

      console.log(`[${i + 1}/${data.length}] Removing background: "${item.title.substring(0, 30)}..."`);
      try {
        // 1. Remove background (accepts URL directly)
        const blob = await removeBackground(imageUrl);

        // 2. Convert Blob to Buffer
        const buffer = Buffer.from(await blob.arrayBuffer());

        // 3. Save to disk
        fs.writeFileSync(localPath, buffer);

        // 4. Update JSON path
        item.image_url = `/bags/${filename}`;
        modifiedCount++;
      } catch (error) {
        console.error(`[!] Error processing item ${i} (${item.title.substring(0, 20)}):`, error.message);
      }
    }
  }

  // Save the JSON back
  if (modifiedCount > 0 || data.some(item => !item.image_url.startsWith("http"))) {
    fs.writeFileSync(JSON_PATH, JSON.stringify(data, null, 2), "utf8");
    console.log(`\n[+] Success! Updated ${modifiedCount} image paths in bags.json.`);
    console.log(`[+] All transparent images are saved under: public/bags/`);
  } else {
    console.log("\n[-] No images were processed.");
  }
}

run().catch(console.error);
